chapter_attention-mechanisms/attention-cues.ipynb

```diff
@@ -0,0 +1,351 @@
+   "outputs": [],
+   "source": [
+    "%load ../utils/plot-utils\n",
+    "%load ../utils/Functions.java"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import ai.djl.ndarray.NDArray;\n",
+    "import ai.djl.ndarray.NDManager;\n",
+    "import ai.djl.ndarray.types.Shape;\n",
+    "import ai.djl.translate.TranslateException;\n",
+    "import tech.tablesaw.plotly.Plot;\n",
+    "import tech.tablesaw.plotly.components.*;\n",
+    "import tech.tablesaw.plotly.traces.HeatmapTrace;\n",
+    "import tech.tablesaw.plotly.traces.Trace;"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "NDManager manager = NDManager.newBaseManager(Functions.tryGpu(0));"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "origin_pos": 4
+   },
+   "source": [
+    "To visualize attention weights,\n",
+    "we define the `showHeatmaps` function.\n",
+    "Its input `matrices` has the shape (number of rows for display, number of columns for display, number of queries, number of keys).\n"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "public static Figure showHeatmaps(\n",
+    "            NDArray matrices,\n",
+    "            String xLabel,\n",
+    "            String yLabel,\n",
+    "            String[] titles,\n",
+    "            int width,\n",
+    "            int height) {\n",
+    "        int numRows = (int) matrices.getShape().get(0);\n",
+    "        int numCols = (int) matrices.getShape().get(1);\n",
+    "\n",
+    "        Trace[] traces = new Trace[numRows * numCols];\n",
+    "        int count = 0;\n",
+    "        for (int i = 0; i < numRows; i++) {\n",
+    "            for (int j = 0; j < numCols; j++) {\n",
+    "                NDArray NDMatrix = matrices.get(i).get(j);\n",
+    "                double[][] matrix =\n",
+    "                        new double[(int) NDMatrix.getShape().get(0)]\n",
+    "                                [(int) NDMatrix.getShape().get(1)];\n",
+    "                Object[] x = new Object[matrix.length];\n",
+    "                Object[] y = new Object[matrix.length];\n",
+    "                for (int k = 0; k < NDMatrix.getShape().get(0); k++) {\n",
+    "                    matrix[k] = Functions.floatToDoubleArray(NDMatrix.get(k).toFloatArray());\n",
+    "                    x[k] = k;\n",
+    "                    y[k] = k;\n",
+    "                }\n",
+    "                HeatmapTrace.HeatmapBuilder builder = HeatmapTrace.builder(x, y, matrix);\n",
+    "                if (titles != null) {\n",
+    "                    builder = (HeatmapTrace.HeatmapBuilder) builder.name(titles[j]);\n",
+    "                }\n",
+    "                traces[count] = builder.build();\n",
+    "            }\n",
+    "            count += 1;\n",
```

Comment:
Shouldn't incrementing count be inside the columns for loop? Alternatively, you can use `traces[count++]` to combine the access and increment

Reply:
Ahh yes you are correct. 

Changed.